src/clj/rems/api/applications_v2.clj

```diff
@@ -0,0 +1,869 @@
+                       :start (:license/start license)
+                       :end (:license/end license)
+                       :approved (:license/accepted license)
+                       :title (:default (:license/title license))
+                       :textcontent (:default (or (:license/link license)
+                                                  (:license/text license)))
+                       :localizations (into {} (for [lang (-> (set (concat (keys (:license/title license))
+                                                                           (keys (:license/link license))
+                                                                           (keys (:license/text license))))
+                                                              (disj :default))]
+                                                 [lang {:title (get-in license [:license/title lang])
+                                                        :textcontent (or (get-in license [:license/link lang])
+                                                                         (get-in license [:license/text lang]))}]))})
+                    (:application/licenses application))
+     :items (map (fn [field]
+                   {:id (:field/id field)
+                    :type (name (:field/type field))
+                    :optional (:field/optional field)
+                    :options (:field/options field)
+                    :maxlength (:field/max-length field)
+                    :value (:field/value field)
+                    :previous-value nil ; TODO
+                    :localizations (into {} (for [lang (set (concat (keys (:field/title field))
+                                                                    (keys (:field/placeholder field))))]
+                                              [lang {:title (get-in field [:field/title lang])
+                                                     :inputprompt (get-in field [:field/placeholder lang])}]))})
+                 (:form/fields application))}))
+
+(defn api-get-application-v1 [user-id application-id]
+  (when-let [v2 (api-get-application-v2 user-id application-id)]
+    (transform-v2-to-v1 v2 user-id)))
+
+;;; v2 API, listing all applications
+
+(defn- applications-view
+  "Projection for the current state of all applications."
+  [applications event]
+  (if-let [app-id (:application/id event)] ; old style events don't have :application/id
+    (update applications app-id application-view event)
+    applications))
+
+(defn- reset-applications-state [_]
+  {:last-processed-event-id 0
+   :applications {}})
+
+(def ^:private applications-state
+  "Stateful projection tracking the current state of all applications.
+  Will be updated periodically by `applications-update-scheduler`,
+  but an update may also be triggered with `trigger-applications-update!`
+  e.g. after committing some new events."
+  (agent (reset-applications-state nil)
```

Comment:
Because `update-applications!` does I/O. An agent ensures that only one update is running in parallel. Atom would allow multiple (slow) updates running in parallel and then it would retry all but one of them. Agents are not optimal either, because they won't discard extra `trigger-applications-update!` invocations. Something like https://github.com/luontola/cqrs-hotel/blob/master/src/main/java/fi/luontola/cqrshotel/framework/util/SingleThreadedTriggerableWorker.java would be best, but AFAIK there is no such built-in concurrency construct so one would need to be created.

Reply:
Or only mark the view as dirty and have the future do the update as soon as required.

Reply:
As discussed lets run the projection every time for now and come back to this later.